### ISSUE TITLE
MEGAFLASH: cartridge and slot 0 flashing overhaul

### ISSUE THREAD

**Issue by lydon42** [MEGA65/mega65-core#683]
With the recent C64 alpha core that supports cartridges, it would be nice if the MEGA65 can boot a specific slot if a cartridge was detected.

This will be continued in #782 

Plan:
- [x] add flag so one slot can flagged as the cartridge slot
  - ~flag can be put in position 32 of the name. This is never displayed.~
  - added as extra flags fields
  - lowest slot with flag wins!
- [x] detect cartridge on startup and select cartridge core slot instead of default slot to boot
  - $D67E & $60 != $60 means ROM cartridge inserted
- [x] add check for `MEGA65-Cart`: Ultimax mode plus `M65` at $8007
  - cartridge RAM can be found at $4008000
- [x] add C128 Cartridge detection
  - define extra core slot marker for possible future C128 core
  - actually implementing real C128 cartridge support might be difficult as EXROM/GAME are not bidirectional on the mainboard
- [x] add secure slot 0 flashing (crc32)
- [x] further secure slot 0 flashing
  - only allow UPGRADE0.COR using core flasher or upgrade0.prg
  - check for MEGA65 core in slot 1
- [x] ~add visual feedback if running core hash != slot 0 core hash?~
  - ~extra field in COR header needed for hash storage?~
  - even with this it is not possible to really determine if we have started the slot 0 flasher... so it can't be done.
- [x] ~~re-add detection for non Ultimax MEGA65 style cartridge (until we know what we are doing...)~~
- [x] add ability to change cartridge core flags and default boot slot in MEGAFLASH
  - [x] add edit screen between main menu and flashing
  - [X] change display of core and flash slot information (currently only slot 0)
  - [X] when flashing a core, display flags from file and allow change before flashing
  - ~~shortcuts for (un)set all?~~
- [x] add bringup flashing via JTAG that does not need user input
- ~~(perhaps) add warning if HICKUP.M65 is found in SD Card root (where?)~~

This branch also handles:
- [x] #684 
- [x] #567 
- [x] #697
- [x] #370 
- [x] #617 
- [x] #625 
- [x] #783

**Comment by lydon42** [comment]
kernel64.src:4114+ does check exactly for CBM80 at $8004, so checking MEGA65 at $8009 seems to be viable.

**Comment by lydon42** [comment]
The `MEGA65` check is not really necessary, we can just do it the C128 way, not lowering /GAME and /EXROM and depending just on a memory signature.
So I will remove this again and rather exchange it with a C128 core detection.

**Comment by lydon42** [comment]
[mega65-tools/dev#6ed6071](https://github.com/MEGA65/mega65-tools/commit/6ed6071233d4f73a6a03b62b895c973fe2f8b006) adds bit2core caps/flags features.

**Comment by lydon42** [comment]
typo in the commit message: 85f387ef7f7929fba41b849a3788b21f08aff582

**Comment by lydon42** [comment]
Current plan:
 - add CRC32 calculation to core integrated MEGAFLASH
 - allow Slot 0 Upgrade from within Core integrated MEGAFLASH

**Comment by lydon42** [comment]
Sucessfull borken slot 0 recovery using etherload to push upgrade0.prg!
Everything needed is in latest build of this branch and mega65-tools development (ethload tool).

**Comment by gardners** [comment]
Examining problems with R3 board.
Things as I discover them:
1. Hardware-accelerated RDID reading does not work. The old soft-ware based method does.
2. mhx_writef() seems to be going crazy sometimes. I can't, for example, reliably use the flash inspector, because it stops writing to the screen.
2.1. The problem seems to be mhx_saddr gets out-of-bounds somehow.
2.2. mhx_posy doesn't reach 25 by the time cursor is at bottom of screen to scroll, so scrolling doesn't happen, and it will advance past $07E8 and potentially start corrupting things.
3. erase_sector() checks sector size using variable addr, not the address_in_sector argument that is passed to it.  This might be causing the selection of the wrong flash sector size to erase, at least some of the time.
4. read_data() uses hardware-accelerated QSPI reading, which seems to not work on R3.

In short, the nexus of the problem seems to be that the R3 QSPI flash chip and the hardware-accelerated QSPI functions in sdcardio.vhdl don't seem to work together.

**Comment by gardners** [comment]
For (2) above I have patches to the flash inspector to reset the mhx cursor to 0,0 which works around the problem.

Need to look more deeply into why the hardware-accelerated QSPI functions are failing on R3.  Likely need to setup a model of the QSPI flash chip under VHDL simulation, and see what is going on.  This will require considerable effort and attention which I may not be able to invest until after the blocking issues for R5/R6 are solved in 0.96, in particular the expansion port issues.

**Comment by lydon42** [comment]
Continued in #782
